File: standalone-container/src/test/scala/com/yahoo/container/standalone/CloudConfigInstallVariablesTest.scala
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.standalone

import com.yahoo.container.standalone.CloudConfigInstallVariables.{toConfigModelsPluginDir, toConfigServer, toConfigServers}
import org.hamcrest.CoreMatchers.is
import org.hamcrest.Matchers.arrayContaining
import org.junit.Assert.assertThat
import org.junit.Test

/**
 * @author Ulf Lilleengen
 * @author Tony Vaagenes
 */
class CloudConfigInstallVariablesTest {

  @Test
  def test_configserver_parsing {
    val parsed = toConfigServers("myhost.mydomain.com")
    assertThat(parsed.length, is(1))
  }

  @Test
  def port_can_be_configured {
    val parsed = toConfigServers("myhost:123")
    val port: Int = parsed(0).port.get()
    assertThat(port, is(123))
  }

  @Test
  def multiple_spaces_are_supported {
    val parsed = toConfigServers("test1     test2")
    assertThat(parsed.size, is(2))

    val hostNames = parsed.map(_.hostName)
    assertThat(hostNames, arrayContaining("test1", "test2"))
  }

  @Test(expected = classOf[IllegalArgumentException])
  def missing_port_gives_exception {
    toConfigServer("myhost:")
  }

  @Test(expected = classOf[IllegalArgumentException])
  def non_numeric_port_gives_exception {
    toConfigServer("myhost:non-numeric")
  }

  @Test
  def string_arrays_are_split_on_spaces {
    val parsed = toConfigModelsPluginDir("/home/vespa/foo /home/vespa/bar ")
    assertThat(parsed.size, is(2))
  }

  @Test
  def string_arrays_are_split_on_comma {
    val parsed = toConfigModelsPluginDir("/home/vespa/foo,/home/vespa/bar,")
    assertThat(parsed.size, is(2))
  }
}

File: standalone-container/src/test/java/com/yahoo/container/standalone/CloudConfigInstallVariablesTest.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.standalone;

import com.yahoo.vespa.model.container.configserver.option.CloudConfigOptions;
import org.junit.Test;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

import static com.yahoo.container.standalone.CloudConfigInstallVariables.toConfigModelsPluginDir;
import static com.yahoo.container.standalone.CloudConfigInstallVariables.toConfigServer;
import static com.yahoo.container.standalone.CloudConfigInstallVariables.toConfigServers;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

/**
 * @author Ulf Lilleengen
 * @author Tony Vaagenes
 */
public class CloudConfigInstallVariablesTest {

    @Test
    public void test_configserver_parsing() {
        CloudConfigOptions.ConfigServer[] parsed = toConfigServers("myhost.mydomain.com");
        assertEquals(1, parsed.length);
    }

    @Test
    public void port_can_be_configured() {
        CloudConfigOptions.ConfigServer[] parsed = toConfigServers("myhost:123");
        int port = parsed[0].port.get();
        assertEquals(123, port);
    }

    @Test
    public void multiple_spaces_are_supported() {
        CloudConfigOptions.ConfigServer[] parsed = toConfigServers("test1     test2");
        assertEquals(2, parsed.length);

        List<String> hostNames = Arrays.stream(parsed).map(cs -> cs.hostName).toList();
        assertTrue(hostNames.containsAll(Arrays.asList("test1", "test2")));
    }

    @Test(expected = IllegalArgumentException.class)
    public void missing_port_gives_exception() {
        toConfigServer("myhost:");
    }

    @Test(expected = IllegalArgumentException.class)
    public void non_numeric_port_gives_exception() {
        toConfigServer("myhost:non-numeric");
    }

    @Test
    public void string_arrays_are_split_on_spaces() {
        String[] parsed = toConfigModelsPluginDir("/home/vespa/foo /home/vespa/bar ");
        assertEquals(2, parsed.length);
    }

    @Test
    public void string_arrays_are_split_on_comma() {
        String[] parsed = toConfigModelsPluginDir("/home/vespa/foo,/home/vespa/bar,");
        assertEquals(2, parsed.length);
    }
}
